nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -848,41 +850,82 @@ impl DataStore {
+    SerdeError(#[from] serde_json::Error),
+
+    #[error("Updated {0} database rows, expected 1")]
+    UnexpectedDatabaseUpdate(usize),
+}
+
+impl DataStore {
+    // See comment for `soft_delete_volume`
+    async fn soft_delete_volume_txn(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
         volume_id: Uuid,
-    ) -> Result<CrucibleResources, Error> {
-        // Grab all the targets that the volume construction request references.
-        // Do this outside the transaction, as the data inside volume doesn't
-        // change and this would simply add to the transaction time.
-        let crucible_targets = {
-            let volume =
-                if let Some(volume) = self.volume_get(volume_id).await? {
-                    volume
-                } else {
-                    // The volume was hard-deleted, return an empty
-                    // CrucibleResources
-                    return Ok(CrucibleResources::V1(
-                        CrucibleResourcesV1::default(),
-                    ));
+        err: OptionalError<SoftDeleteTransactionError>,
+    ) -> Result<CrucibleResources, diesel::result::Error> {
+        // Grab the volume, and check if the volume was already soft-deleted.
+        // We have to guard against the case where this function is called
+        // multiple times, and that is done by soft-deleting the volume during
+        // the transaction, and returning the previously serialized list of
+        // resources to clean up.
+        let volume = {
+            use db::schema::volume::dsl;
+
+            let volume = dsl::volume
+                .filter(dsl::id.eq(volume_id))
+                .select(Volume::as_select())
+                .get_result_async(conn)
+                .await
+                .optional()?;
+
+            let volume = if let Some(v) = volume {
+                v
+            } else {
+                // The volume was hard-deleted
+                return Ok(CrucibleResources::V1(
+                    CrucibleResourcesV1::default(),
+                ));
+            };
+
+            if volume.time_deleted.is_some() {
+                // this volume was already soft-deleted, return the existing
+                // serialized CrucibleResources
+                let existing_resources = match volume
+                    .resources_to_clean_up
+                    .as_ref()
+                {
+                    Some(v) => serde_json::from_str(v)
+                        .map_err(|e| err.bail(e.into()))?,
+
+                    None => {
+                        // Even volumes with nothing to clean up should have a
+                        // serialized CrucibleResources that contains empty
+                        // vectors instead of None. Instead of panicing here
+                        // though, just return the default (nothing to clean
+                        // up).
+                        CrucibleResources::V1(CrucibleResourcesV1::default())
```

Comment:
For this and the case above where we make a default crucible resource, should we just return an `Option` and bail out early?

Reply:
That's possible yeah. I'm inclined to pass around these empty vectors to avoid having to `if let Some` all over though.